File: scala/src/main/scala/clase6/ADTUnbundledDeclarativeSecure.scala
package clase6

object ADTUnbundledDeclarativeSecure extends App {

  private case class Key()

  case class Wrapper[T](private val content: List[T]) {
    def unwrap(k: Key): List[T] = content
  }

  def stack[T](): Wrapper[T] = Wrapper[T](Nil)

  def push[T](stack: Wrapper[T], element: T) = Wrapper[T](element :: stack.unwrap(Key()))

  def pop[T](s: Wrapper[T]): (T, Wrapper[T]) = {
    s.unwrap(Key()) match {
      case head :: tail => (head, Wrapper[T](tail))
    }
  }

  def isEmpty[T](s: Wrapper[T]) = s.unwrap(Key()) == Nil

  assert(isEmpty(stack()))

  val s = push(push(stack(), "ariel"), "alejandro")

  assert(!isEmpty(s))

  assert(pop(s)._1 == "alejandro")
  assert(pop(pop(s)._2)._1 == "ariel")
  assert(isEmpty(pop(pop(s)._2)._2))

}

File: haskell/src/clase6/ADTUnbundledDeclarativeSecure.hs
module ADTUnbundledDeclarativeSecure(Wrapper, stack, push, pop, isEmpty) where

-- private type
data Key = Key

data Wrapper t = Wrapper {
  -- data is actually stored in the closure's environment
  content :: Key -> [t]
}

wrap value = Wrapper (\k -> value)

unwrap wrapper key = ((content wrapper) key)

stack = wrap []

push stack element = wrap (element : (unwrap stack Key))

doPop (head : tail) = (head, wrap tail)
pop stack = doPop (unwrap stack Key)

isEmpty stack = checkEmpty (unwrap stack Key)
checkEmpty [] = True
checkEmpty stack = False

main = let s = push (push stack "ariel") "alejandro" in do
    print (isEmpty stack) -- True
    print (isEmpty s) -- False
    print (fst (pop (s))) -- "alejandro"
    print (fst (pop (snd (pop (s))))) -- "ariel"
    print (isEmpty (snd (pop (snd (pop (s)))))) -- True
